### PR TITLE
[#32] Remove or annotate PartialEq and Eq impls for constant time

### PR DESCRIPTION
**Pull request by clintfred**
For the internal types, if `PartialEq` and `Eq` were not needed for anything but the tests, they were moved to the tests or will only compile with the test cfg. This was the majority of the changes. If we decide we want them back, we could put them under a feature flag or add them back for non-secret values.

Some `PartialEq` instances are necessary for (already marked) non-constant-time functionality. These were marked as such. Signature and AuthHashes require equality (but are not secret) and were also annotated as such.

For the public API, all non-secret values still have `PartialEq`. Secret values (`Plaintext`, `DerivedSymmetricKey`, `PrivateKey`) have `PartialEq` using a `Revealed` wrapper type. So (non-constant-time) equality is there, but it will be obvious that the values are revealed by an equality check.

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -55,7 +56,14 @@ impl<CR: rand::CryptoRng + rand::RngCore> Api<Sha256, Ed25519, RandomBytes<CR>>
 }
 
 /// Hashed but not encrypted Plaintext used for envelope encryption
-new_bytes_type!(DerivedSymmetricKey, 32);
+/// If you are looking for PartialEq for DerivedSymmetricKey, see PartialEq for Revealed<DerivedSymmetricKey>
+new_bytes_type_no_eq!(DerivedSymmetricKey, 32);
+
+impl PartialEq for Revealed<DerivedSymmetricKey> {
+    fn eq(&self, other: &Revealed<DerivedSymmetricKey>) -> bool {
+        self.0.bytes == other.0.bytes
```
[the remark this answers is not in the document]
Comment (clintfred) [reply src/api.rs]: It is not; thus the `Revealed` wrapper. Both values would be revealed by calling this.

File: src/api.rs [hunk 2]
```diff
@@ -960,7 +995,15 @@ impl PublicKey {
...
 
-#[derive(Eq, PartialEq, Default, Debug)]
+impl PartialEq for PublicKey {
+    fn eq(&self, other: &PublicKey) -> bool {
+        self.x == other.x && self.y == other.y
+    }
+}
+
+#[derive(Default, Debug)]
+#[cfg_attr(test, derive(PartialEq))]
```
Comment (coltfred) [inline src/api.rs]: Seems like getting rid of these cfg_attr and using Revealed in the tests would be better.

File: src/api.rs [hunk 3]
```diff
@@ -1032,6 +1081,9 @@ pub(crate) mod test {
     use hex;
     use rand_chacha;
 
+    // just for tests...
+    _bytes_eq!(DerivedSymmetricKey);
```
Comment (coltfred) [inline src/api.rs]: Maybe test using Revealed?

File: src/internal/ed25519.rs [hunk 4]
```diff
@@ -193,6 +186,14 @@ pub(crate) mod test {
     use crate::internal::array_concat_32;
     use ed25519_dalek::SecretKey;
 
+    impl PartialEq for SigningKeypair {
```
Comment (coltfred) [inline src/internal/ed25519.rs]: Seems like you might want to have a Revealed implementation for this as well?

File: src/internal/homogeneouspoint.rs [hunk 5]
```diff
@@ -37,7 +38,12 @@ pub struct HomogeneousPoint<T> {
 }
 
 impl<T: One + Field + From<u32> + Hashable> HomogeneousPoint<T> {
-    pub fn from_x_y((x, y): (T, T)) -> Result<HomogeneousPoint<T>, PointErr> {
+    /// Construct a HomogeneousPoint from an x and y portion.
+    ///
+    /// Revealed - `x` and `y` values
+    pub fn from_x_y(
+        (Revealed(x), Revealed(y)): (Revealed<T>, Revealed<T>),
```
Comment (coltfred) [inline src/internal/homogeneouspoint.rs]: This seems unrelated to the issue we're working on.

File: src/internal/homogeneouspoint.rs [hunk 6]
```diff
@@ -498,6 +452,59 @@ pub mod test {
...
     use hex;
     use proptest::prelude::*;
+
+    impl<T> PartialEq for HomogeneousPoint<T>
+    where
+        T: Field,
+    {
+        fn eq(&self, other: &HomogeneousPoint<T>) -> bool {
+            match (*self, *other) {
+                (ref p1, ref p2) if p1.is_zero() && p2.is_zero() => true,
```
Comment (coltfred) [inline src/internal/homogeneouspoint.rs]: You can remove these 2 lines, they're optimization that's especially unimportant in tests.

File: src/internal/homogeneouspoint.rs [hunk 7]
```diff
@@ -498,6 +452,59 @@ pub mod test {
...
+
+    impl<T> Eq for HomogeneousPoint<T> where T: Field {}
+
+    impl<T> PartialEq for TwistedHPoint<T>
+    where
+        T: ExtensionField,
+    {
+        fn eq(&self, other: &TwistedHPoint<T>) -> bool {
+            match (*self, *other) {
+                (ref p1, ref p2) if p1.is_zero() && p2.is_zero() => true,
```
Comment (coltfred) [inline src/internal/homogeneouspoint.rs]: You can remove these 2 lines, they're optimization that's especially unimportant in tests.

File: src/internal/mod.rs [hunk 8]
```diff
@@ -187,7 +194,8 @@ impl AuthHash {
     }
 }
 
-#[derive(Eq, PartialEq, Clone, Debug)]
+#[derive(Clone, Debug)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
```
Comment (coltfred) [inline src/internal/mod.rs]: Above you left the Eq for AuthHash because it's not secret, but changed it here. I'm confused about why.

File: src/api.rs (hunk 2 above)
Comment (clintfred) [reply src/api.rs]: We could do that. We would lose the equality checks on the `_intenal_key` member, which I think we should probably maintain.

Comment (clintfred) [reply src/api.rs]: We could do that. We would lose the equality checks on the `_intenal_key` member, which I think we should probably maintain.

File: src/internal/ed25519.rs (hunk 4 above)
Comment (clintfred) [reply src/internal/ed25519.rs]: Yep. Good catch.

File: src/internal/mod.rs (hunk 8 above)
Comment (clintfred) [reply src/internal/mod.rs]: AuthHash does require equality for checking [if hashes are the same](https://github.com/IronCoreLabs/recrypt-rs/blob/a4e02119b7998e85bdea48fdef27166ef4f542da/src/internal/mod.rs#L592).  There is currently no code that checks equality on `EncryptedValue`s
